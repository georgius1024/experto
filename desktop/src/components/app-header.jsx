import React, { memo } from 'react'
import Navbar from 'react-bootstrap/Navbar'
import NavDropdown from 'react-bootstrap/NavDropdown'
import Nav from 'react-bootstrap/Nav'
import { withRouter } from 'react-router'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import actions from '../store/actions'

function AppHeader({ authenticated, history, expertName, unregister }) {
  function logout() {
    unregister()
    history.push('/')
  }
  function goMain(e) {
    e.preventDefault()
    history.push('/')
  }
  const brand = (
    <Navbar.Brand href="#" onClick={goMain}>
      Experto Crede
    </Navbar.Brand>
  )
  if (!authenticated) {
    return (
      <Navbar variant="dark" bg="primary" sticky="top">
        {brand}
        <Navbar.Text className="ml-auto mr-2">Эксперт не зарегистрирован</Navbar.Text>
      </Navbar>
    )
  } else {
    return (
      <Navbar variant="dark" bg="primary" sticky="top" expand="md" className="app-bar">
        {brand}
        <Navbar.Toggle aria-controls="basic-navbar-nav" />
        <Navbar.Collapse id="basic-navbar-nav">
          <Nav className="ml-auto mr-2">
            <NavDropdown title={expertName} className="private-section">
              <NavDropdown.Item onClick={logout}>Выйти</NavDropdown.Item>
            </NavDropdown>{' '}
          </Nav>
        </Navbar.Collapse>
      </Navbar>
    )
  }
}

AppHeader.propTypes = {
  authenticated: PropTypes.bool,
  expertName: PropTypes.string,
  history: PropTypes.object
}

const mapStateToProps = state => {
  return {
    authenticated: Boolean(state.registration.accessToken),
    expertName: state.registration.name
  }
}

const mapDispatchToProps = dispatch => {
  return {
    unregister: () => dispatch(actions.unregister())
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(memo(AppHeader)))
