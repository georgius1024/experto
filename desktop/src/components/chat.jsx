import React, { memo, useState } from 'react'
import ListGroup from 'react-bootstrap/ListGroup'
import Form from 'react-bootstrap/Form'
import moment from 'moment-timezone'

function Chat({ messages, onPostMessage }) {
  const [message, setMessage] = useState('')
  function onSubmit(event) {
    event.preventDefault()
    onPostMessage(message)
    setMessage('')
  }
  function onChange(event) {
    setMessage(event.target.value)
  }

  const history = messages.map(({ message, from, at }) => {
    const time = moment(at).format('HH:mm:ss')
    return (
      <ListGroup.Item key={time}>
        <p>
          <b>
            {time} - {from}
          </b>
        </p>
        <p>{message}</p>
      </ListGroup.Item>
    )
  })
  const form = (
    <Form onSubmit={onSubmit}>
      <Form.Group>
        <Form.Control type="text" placeholder="Введите сообщение" value={message} onChange={onChange} />
      </Form.Group>
    </Form>
  )
  return (
    <>
      <ListGroup>{history}</ListGroup>
      {form}
    </>
  )
}

export default memo(Chat)
