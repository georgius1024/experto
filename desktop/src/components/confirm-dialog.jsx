import React, { memo } from 'react'
import PropTypes from 'prop-types'
import Modal from 'react-bootstrap/Modal'
import Button from 'react-bootstrap/Button'
function ConfirmDialog({ active, caption, body, onConfirm, onCancel, confirmButton, cancelButton }) {
  const hasConfirmButton = Boolean(onConfirm)
  const hasCancelButton = Boolean(onCancel)
  return (
    <Modal show={active} backdrop="static" centered>
      {caption && (
        <Modal.Header>
          <Modal.Title>{caption}</Modal.Title>
        </Modal.Header>
      )}
      <Modal.Body>{body}</Modal.Body>
      <Modal.Footer>
        {hasConfirmButton && (
          <Button variant="primary" onClick={onConfirm} className="w-10em">
            {confirmButton || 'Подтвердить'}
          </Button>
        )}
        {hasCancelButton && (
          <Button variant="secondary" onClick={onCancel} className="w-10em">
            {cancelButton || 'Отмена'}
          </Button>
        )}
      </Modal.Footer>
    </Modal>
  )
}

ConfirmDialog.propTypes = {
  active: PropTypes.bool,
  caption: PropTypes.string,
  body: PropTypes.string, 
  onConfirm: PropTypes.func, 
  onCancel: PropTypes.func, 
  confirmButton: PropTypes.string,
  cancelButton: PropTypes.string,
}

export default memo(ConfirmDialog)
