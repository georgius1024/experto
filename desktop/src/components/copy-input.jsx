import React, { memo } from 'react'
import copyToClipboard from '../utils/copy-to-clipboard'
import Form from 'react-bootstrap/Form'
import Button from 'react-bootstrap/Button'

export default memo(function({ value }) {
  const style = { width: 'calc(100% - 60px)' }
  return (
    <>
      <Button variant="link" onClick={copyToClipboard(value)} className="float-right">
        <i className="fa fa-copy" />
      </Button>
      <Form.Control readOnly={true} value={value} plaintext={true} style={style} />
    </>
  )
})
