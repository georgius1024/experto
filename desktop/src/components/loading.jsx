/*
 Show transparent layer with wait cursor
*/
import React, { useEffect, useState } from 'react'
import Api from '../api'
import styles from './loading.module.scss'

function LoadingComponent(props) {
  const [loading, setLoading] = useState(false)
  useEffect(() => {
    const subscription = Api.busy$.subscribe(busy => {
      setLoading(busy)
      if (busy) {
        setTimeout(() => {
          setLoading(false)
        }, 5000)
      }
    })
    return function cleanup() {
      subscription.unsubscrube()
    }
  }, [])
  return loading && <div className={styles.loading} />
}
export default LoadingComponent
