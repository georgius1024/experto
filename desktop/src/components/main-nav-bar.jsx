import React, { memo } from 'react'
import PropTypes from 'prop-types'
import ButtonToolbar from 'react-bootstrap/ButtonToolbar'
import ButtonGroup from 'react-bootstrap/ButtonGroup'
import Button from 'react-bootstrap/Button'
import Dropdown from 'react-bootstrap/Dropdown'
import styles from './main-nav-bar.module.scss'
function NavigationBar({ rooms, selectedRoom, onSelect, onCreate, onUpdate, onDelete }) {
  if (!rooms.length) {
    return null
  }
  const selectedId = () => selectedRoom && selectedRoom._id

  const currentRoomIndex = () => {
    return rooms.findIndex(e => e._id === selectedId())
  }

  const canMoveNext = () => {
    return currentRoomIndex() < rooms.length - 1
  }

  const onMoveNext = () => {
    if (canMoveNext()) {
      onSelect(rooms[currentRoomIndex() + 1])
    }
  }

  const canMovePrev = () => {
    return currentRoomIndex() > 0
  }

  const onMovePrev = () => {
    if (canMovePrev()) {
      onSelect(rooms[currentRoomIndex() - 1])
    }
  }
  return (
    <ButtonToolbar>
      <Dropdown as={ButtonGroup} className="mr-2">
        <Button onClick={onMovePrev} disabled={!canMovePrev()}>
          <i className="fa fa-caret-left" />
        </Button>
        <Button>{`${currentRoomIndex() + 1} / ${rooms.length}`}</Button>
        <Dropdown.Toggle split className={styles.split} />
        <Dropdown.Menu>
          {rooms.map(room => {
            const select = e => {
              e.preventDefault()
              onSelect(room)
            }
            return (
              <Dropdown.Item href="#" onClick={select} key={room._id}>
                {room.roomName}
              </Dropdown.Item>
            )
          })}
        </Dropdown.Menu>
        <Button onClick={onMoveNext} disabled={!canMoveNext()}>
          <i className="fa fa-caret-right" />
        </Button>
      </Dropdown>
      <ButtonGroup className="mr-2">
        <Button variant="primary" className="w-10em" onClick={onCreate}>
          Создать
        </Button>
        <Button variant="primary" className="w-10em" onClick={onUpdate}>
          Редактировать
        </Button>
        <Button variant="danger" className="w-10em" onClick={onDelete}>
          Удалить
        </Button>
      </ButtonGroup>
    </ButtonToolbar>
  )
}

NavigationBar.propTypes = {
  rooms: PropTypes.array,
  selectedRoom: PropTypes.object,
  onSelect: PropTypes.func,
  onCreate: PropTypes.func,
  onUpdate: PropTypes.func,
  onDelete: PropTypes.func
}

export default memo(NavigationBar)
