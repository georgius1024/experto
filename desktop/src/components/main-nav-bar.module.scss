.split {
  border-radius: 0 !important;
}
