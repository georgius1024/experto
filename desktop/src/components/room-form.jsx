import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Prompt } from 'react-router'
import Form from 'react-bootstrap/Form'
import Button from 'react-bootstrap/Button'

const roomModel = {
  _id: 'new',
  roomName: '',
  listenerName: '',
  listenerEmail: '',
  date: ''
}

class RoomForm extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      initial: roomModel,
      model: roomModel,
      errors: {},
      errorCount: 0
    }
    this.onChange = this.onChange.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
    this.onCancel = this.onCancel.bind(this)
    this.formIsDirty = this.formIsDirty.bind(this)
  }

  componentDidMount() {
    const newModel = { ...roomModel, ...this.props.model }
    this.setState({
      initial: newModel,
      model: newModel
    })
  }

  componentDidUpdate(prevProps) {
    if (prevProps.model !== this.props.model) {
      const newModel = { ...roomModel, ...this.props.model }
      this.setState({
        initial: newModel,
        model: newModel
      })
    }
  }
  formIsDirty() {
    let dirty = false
    Object.keys(this.state.model).forEach(field => {
      if (this.state.model[field] !== this.state.initial[field]) {
        dirty = true
      }
    })
    return dirty
  }

  onChange(event) {
    const field = event.target.name
    const value = event.target.value
    const model = { ...this.state.model }
    model[field] = value
    this.setState(
      {
        model: model
      },
      () => this.validate(field)
    )
  }

  valid(field) {
    const value = this.state.model[field]
    if (!value) {
      return 'Обязательное значение'
    }
    switch (field) {
    case 'listenerEmail':
      if (
        !/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/.test(
          value
        )
      ) {
        return 'Введите правильный e-mail'
      }
      break
    default:
    }
    return true
  }

  validate(field) {
    const valid = this.valid(field)
    const errors = { ...this.state.errors }
    if (valid === true) {
      delete errors[field]
    } else {
      errors[field] = valid
    }
    this.setState({
      errors,
      errorCount: Object.keys(errors).length
    })
  }

  validateAll() {
    const errors = {}
    Object.keys(this.state.model).forEach(field => {
      const valid = this.valid(field)
      if (valid !== true) {
        errors[field] = valid
      }
    })
    const errorCount = Object.keys(errors).length
    this.setState({
      errors,
      errorCount
    })
    return errorCount
  }

  onSubmit(event) {
    event.preventDefault()
    if (this.validateAll() === 0) {
      this.setState({
        initial: { ...this.state.model }
      })
      this.props.onSubmit(this.state.model)
    }
  }

  onCancel() {
    this.props.onCancel()
  }

  render() {
    const model = this.state.model
    const errors = this.state.errors
    const errorCount = this.state.errorCount
    const readOnly = this.props.readOnly
    const plainText = this.props.plainText
    const hasSubmitButton = Boolean(this.props.onSubmit)
    const hasCancelButton = Boolean(this.props.onCancel)
    const formIsDirty = this.formIsDirty()
    return (
      <Form onSubmit={this.onSubmit}>
        <Form.Group>
          <Form.Label>Время встречи</Form.Label>
          <Form.Control
            type="datetime-local"
            name="date"
            plaintext={plainText}
            readOnly={readOnly}
            value={model['date']}
            onChange={this.onChange}
            placeholder="Выберите время встречи"
            required
          />
          <div className="error-text">{errors['date']}</div>
        </Form.Group>
        <Form.Group>
          <Form.Label>Название комнаты</Form.Label>
          <Form.Control
            name="roomName"
            value={model['roomName']}
            plaintext={plainText}
            readOnly={readOnly}
            onChange={this.onChange}
            type="text"
            maxLength="40"
            placeholder="Введите название"
            required
          />
          <div className="error-text">{errors['roomName']}</div>
        </Form.Group>

        <Form.Group>
          <Form.Label>Слушатель</Form.Label>
          <Form.Control
            name="listenerName"
            value={model['listenerName']}
            plaintext={plainText}
            readOnly={readOnly}
            onChange={this.onChange}
            type="text"
            maxLength="40"
            placeholder="Введите ФИО"
            required
          />
          <div className="error-text">{errors['listenerName']}</div>
        </Form.Group>

        <Form.Group>
          <Form.Label>Email слушателя</Form.Label>
          <Form.Control
            name="listenerEmail"
            value={model['listenerEmail']}
            plaintext={plainText}
            readOnly={readOnly}
            onChange={this.onChange}
            type="email"
            placeholder="Введите email"
            required
          />
          <div className="error-text">{errors['listenerEmail']}</div>
        </Form.Group>
        <div className="text-right">
          {hasSubmitButton && (
            <Button variant="primary" disabled={errorCount} type="submit" className="w-10em mr-4">
              Сохранить
            </Button>
          )}
          {hasCancelButton && (
            <Button variant="secondary" onClick={this.onCancel} type="button" className="w-10em">
              Отмена
            </Button>
          )}
        </div>
        <Prompt when={formIsDirty} message="Точно-точно, прямо точно?" />
      </Form>
    )
  }
}

RoomForm.propTypes = {
  model: PropTypes.object,
  readOnly: PropTypes.bool,
  plainText: PropTypes.bool,
  onSubmit: PropTypes.any,
  onCancel: PropTypes.any
}

export { RoomForm, roomModel }
