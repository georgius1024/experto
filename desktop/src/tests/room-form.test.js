import React from 'react'
import renderer from 'react-test-renderer'
import { RoomForm as Component } from '../components/room-form'
import { MemoryRouter } from 'react-router-dom'
describe('Room-form component', () => {
  test('it matches the snapshot', () => {
    const component = renderer.create(
      <MemoryRouter>
        <Component />
      </MemoryRouter>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})
