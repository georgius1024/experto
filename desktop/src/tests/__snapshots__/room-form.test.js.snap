// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room-form component it matches the snapshot 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className="form-label"
    >
      Время встречи
    </label>
    <input
      className="form-control"
      name="date"
      onChange={[Function]}
      placeholder="Выберите время встречи"
      required={true}
      type="datetime-local"
      value=""
    />
    <div
      className="error-text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="form-label"
    >
      Название комнаты
    </label>
    <input
      className="form-control"
      maxLength="40"
      name="roomName"
      onChange={[Function]}
      placeholder="Введите название"
      required={true}
      type="text"
      value=""
    />
    <div
      className="error-text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="form-label"
    >
      Слушатель
    </label>
    <input
      className="form-control"
      maxLength="40"
      name="listenerName"
      onChange={[Function]}
      placeholder="Введите ФИО"
      required={true}
      type="text"
      value=""
    />
    <div
      className="error-text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="form-label"
    >
      Email слушателя
    </label>
    <input
      className="form-control"
      name="listenerEmail"
      onChange={[Function]}
      placeholder="Введите email"
      required={true}
      type="email"
      value=""
    />
    <div
      className="error-text"
    />
  </div>
  <div
    className="text-right"
  />
</form>
`;
